src/python_be.cc

```diff
@@ -1468,6 +1494,80 @@ ModelInstanceState::ProcessRequests(
   return;
 }
 
+void
+ModelInstanceState::SendBLSDecoupledResponse(
+    std::unique_ptr<InferResponse> response_ptr)
+{
+  bool is_response_batch_set = false;
+  ResponseBatch* response_batch;
+  std::unique_ptr<PbString> pb_error_message;
+  std::unique_ptr<IPCMessage> ipc_message;
+  AllocatedSharedMemory<char> response_batch_shm;
+
+  try {
+    ipc_message =
+        IPCMessage::Create(Stub()->ShmPool(), true /* inline_response */);
+    response_batch_shm = Stub()->ShmPool()->Construct<char>(
+        sizeof(ResponseBatch) + sizeof(bi::managed_external_buffer::handle_t));
+    response_batch =
+        reinterpret_cast<ResponseBatch*>(response_batch_shm.data_.get());
+    bi::managed_external_buffer::handle_t* response_handle =
+        reinterpret_cast<bi::managed_external_buffer::handle_t*>(
+            response_batch_shm.data_.get() + sizeof(ResponseBatch));
+    ipc_message->Args() = response_batch_shm.handle_;
+
+    response_batch->batch_size = 1;
+    response_batch->has_error = false;
+    response_batch->is_error_set = false;
+    response_batch->cleanup = false;
+    response_batch->response_size = 1;
+    is_response_batch_set = true;
+
+    response_ptr->SaveToSharedMemory(Stub()->ShmPool());
+    for (auto& output_tensor : response_ptr->OutputTensors()) {
+      // For GPU tensors we need to store the memory release id in
+      // memory manager.
+      if (!output_tensor->IsCPU()) {
+#ifdef TRITON_ENABLE_GPU
+        std::unique_ptr<MemoryRecord> gpu_memory_record =
+            std::make_unique<GPUMemoryRecord>(
+                output_tensor->Memory()->DataPtr());
+        uint64_t memory_release_id =
+            Stub()->GetMemoryManager()->AddRecord(std::move(gpu_memory_record));
+        output_tensor->Memory()->SetMemoryReleaseId(memory_release_id);
+#endif
+      }
+    }
+
+    *response_handle = response_ptr->ShmHandle();
+  }
+  catch (const PythonBackendException& pb_exception) {
+    if (is_response_batch_set) {
+      response_batch->has_error = true;
+      LOG_IF_EXCEPTION(
+          pb_error_message =
+              PbString::Create(Stub()->ShmPool(), pb_exception.what()));
+
+      if (pb_error_message != nullptr) {
+        response_batch->is_error_set = true;
+        response_batch->error = pb_error_message->ShmHandle();
+      }
+    } else {
+      LOG_MESSAGE(TRITONSERVER_LOG_ERROR, pb_exception.what());
+    }
+  }
+
+  {
+    bi::scoped_lock<bi::interprocess_mutex> lock{
+        *(ipc_message->ResponseMutex())};
+    bool success = false;
+    while (!success) {
+      Stub()->BLSResponseQueue()->Push(ipc_message->ShmHandle(), 1000, success);
+    }
+    ipc_message->ResponseCondition()->wait(lock);
```

Comment:
This also needs to be wrapped around in a while loop.

Reply:
Added.